### ISSUE TITLE
Renew expired keys

### ISSUE THREAD

**Issue by ayoyoness** [pixelated/pixelated-user-agent#815]
**Narrative:**
As a user with expired keys I want my keys renewed so that I am able to use secure email communication: encypt/decrypt, sign/verify emails.

**ACs:**
- the new key pair are signed by my old key pair
- old emails can be read
- the first email to any users should have the new public key attached
- users with leap providers should automatically be able to sign/encrypt emails to the user,  after the first email from her/him.
- users from non-leap providers will need to refresh the user's public key.

**Comment by ayoyoness** [comment]
Tasks:
[] implement extend_key functionality in python-gnupg
[] add corresponding wrapper in keymanager
[] invoke the functionality in the login process of the user-agent

**Comment by thaissiqueira** [comment]
We also should deal with who are trying to send an email to someone that had your key expired.
I'm not sure if is safe say to sender that the recipient's key expired.
@SweetVirginia, some thought?

**Comment by bwagnerr** [comment]
@thaissiqueira I think the question arrives even before that, if a recipient has no valid keys (only an expired one), should it receive the email in plain text?

**Comment by ayoyoness** [comment]
The first email that goes out to a particular contact that had already sent you an encrypted email, will receive the extended key as an attachment. 
People using leap providers will automatically import this key and therefore will have the new expiry date associated with that public key in the their keyring.
People not on leap providers will have to manually import it (at least thunderbird will show something when a key is attached, incentivising the user to do something).

Sending an email to someone who has your expired key (or equivalently, receiving an email from the user) does not require his keys so encryption will still be okay... signature would fail though. But when the attached extended key is imported, all should be restored to their normal behaviour.

Someone sending an email to you will not work if the imported public key (of the user) is expired.  We need to work around this: either we leave it as is and feedback "email not sent due du expired key" or send the email in plain text (with some warnings or UX component).

**Comment by thaissiqueira** [comment]
> @thaissiqueira I think the question arrives even before that, if a recipient has no valid keys (only an expired one), should it receive the email in plain text?

Is it a new issue?

**Comment by deniscostadsc** [comment]
@ayoyoness you did PR to python GnuPG, right? Did you do it against @isis repo or some other repo? I'm asking because, if was on @isis, we are going ot get stuck as Zara PRs got stuck.

**Comment by ayoyoness** [comment]
update on this:
- python-gnupg implementation (with PR -- see #818 ) is done
- corresponding call in keymanager (where keys are actually unlocked and imported) is underway: refactoring needed.
- remaining, not started: invoking  key extension from login in the user-agent

**Comment by ayoyoness** [comment]
@deniscostadsc yes, the PR is in isis repo... but:
1. there is already movement in there... merging will happen soon -- as in not months ^^
2. as agreed, we will use the branch for the time being.

**Comment by ayoyoness** [comment]
@olabini 

Do you think we should:
1. finish the key extension and
2. extend the current key and 
3. generate a new key and
4. sign the  new key with the old one.

Or we should stop at 1. That is, only extend the key.

The reason, I'm asking is to make sure that:
1. Everyone is okay with the approach of just extending the current key... given that we set out to do the recommend approach: generate a new key and sign it with old key. However, our challenge is that we could not sign any key with an expired key... so then we're off-course and back at the less recommend approach: just extend the key... of course now that we have a non-expired key (after extension), we can technically revert back to the recommended approach.

2. We have fully considered all possibilities on how to sign a key with an expired key... my gpg knowledge is unfortunately very limited :)

Comments from anyone welcome as well
@kalikaneko @bwagnerr

**Comment by olabini** [comment]
Can you expand on this? I really don't understand what you're asking here.

**Comment by ayoyoness** [comment]
Our main goal is to allow people who have or will have expired keys, to have access and use their account normally. 
Now, we have agreed that the best approach for this is to generate a new key for the user and sign it with his old (expired) key. The problem is that gpg does not allow signing from an expired  key. So then, we had to extend the expired key. Now we don't have an expired key anymore. 
The question at this point is, should we stop here (old key being extended) as the user already can resume using the system? 
Or should we still generate a new key and sign it the extended-key, as that is what we have originally established as best practice and aimed for? (FYI, I have started implementing key signing in python-gnupg)
Or is there a way to sign from an expired key in gpg that I might have overlooked? this will by-pass the whole key expiry extension part (as was our original approach).

**Comment by olabini** [comment]
You could try adding the --expert option and see if that allows you signing using an expired key, but otherwise I don't think it's possible.
I personally think it's better to just do the right thing here. It doesn't seem like that should take much time to implement anyway.

**Comment by zariye** [comment]
@ayoyoness created this [ticket](https://leap.se/code/issues/8648) in case we need to discuss something with azul about fetching keys from another leap domain

**Comment by deniscostadsc** [comment]
Ola pointed out that don't support decryption with two keys is a bug. Work on that isn't going to divert too much from initial goal? Because, as far as I got, we are going to implement that, right?

**Comment by bwagnerr** [comment]
Hey @zariye @ayoyoness 
Aren't these clearly two steps that can deliver value for themselves?
Shouldn't we finish the "Extend expired keys", by making pixelated use it, before we start working on "Renew expired keys"?

I fear we are losing focus of doing the smallest piece to make sure people on wazokazi recover their accounts that have the key expired

**Comment by zariye** [comment]
@deniscostadsc exactly, we will create a ticket for that after we have finished this story and we will also figure out how to aces keys  from another leap-provider

@bwagnerr i just renamed the story, bc i knew it is kinda confusing. The idea here is to extend the key only in order to create a new key and sign it by the old one (leap validation rule). That is how we understood Ola, when he said:

> I personally think it's better to just do the right thing here. It doesn't seem like that should take much time to implement anyway.

**Comment by ayoyoness** [comment]
So this is done within the adjusted story narrative and acceptance criteria above.
It is in the branch _key_expiry_ 
The two next steps are in #850 and #851

**Comment by varac** [comment]
I won't work on this today and tomorrow is my day of so I'm moving this back to ready-for-review.

**Comment by deniscostadsc** [comment]
@zariye @ayoyoness

I' not sure how can I review this issue. Can you explain a bit a small scenario to test what as done on this issue.

**Comment by deniscostadsc** [comment]
All the work has been done on the branch key_expiry, right? But this branch is 5 commits ahead, 43 commits behind master. Is this supposed to be like that? All the work is going to be done on a branch and then merged in the end?

**Comment by bwagnerr** [comment]
I can't review this issue before #859 and #855 are merged (on our fork on #855 case and on python-gnupg on #859 case).

It looks like we also shouldn't merge the branch key_expiry while the PRs are pending (because the changes will break without the earlier PRs)

I'll block this issue until we figure out what to do with the PRs.

Also a question, considering we have 5 PRs to python-gnupg (going back to july this year), should we just fork it, incorporate our changes and go on with the issues?

**Comment by ayoyoness** [comment]
I think we have always had in mind from the beginning that we won't depend on the PRs and would use our branches in the meantime. That is both for bitmask-dev/keymanager PRs and python-gnupg PRs.

I have made the changes to that effect: use our branch of bitmask-dev, python-gnupg  and merge key_expiry to master in the user-agent... I just have not pushed yet, as @varac  wanted to start review today and i didn't want it to hit wazokazi  yet...

**Comment by ayoyoness** [comment]
So I just pushed on key_expiry branch in the UA
before any review, whenever that happens, we should merge that branch into master and review it from there.

**Comment by varac** [comment]
@bwagnerr please go ahead if you like to review!

**Comment by bwagnerr** [comment]
Hey @ayoyoness @zariye , I like the idea of not depending on the PRs, but only if we have the changes in our fork (that is what we use anyway).

I don't feel comfortable reviewing this with local customization of 3 components (python-gnupg, bitmask-dev and pixelated-user-agent).

Do you think we should merge these changes to our forks develop so we can test the user agent properly with this on master?

Also we don't have a fork of gnupg, in that case we need start a fork of it (and use that fork on the UA), so we can merge the changes and review/use them

**Comment by ayoyoness** [comment]
I'm fine either way. 

The reason why I'm proposing to use a different branch (https://0xacab.org/pixelated/bitmask-dev/tree/feat_fetch_remote_on_expiry) than develop on our fork of bitmask-dev is that  it is easier to manage if we keep that inline with leap_gitlab develop... and we keep the branch (`feat_fetch_remote_on_expiry`) also inline with leap_gitlab develop.

That branch `feat_fetch_remote_on_expiry` already uses a fork of python-gnupg

But again, I'm completely fine either way. I will leave it up to you.
Please remember to merge with pixelated-user-agent/master  and review from there.

**Comment by thaissiqueira** [comment]
Now we need to merge the branch with the changes to support key renewing.
Also we need to merge our PRs  to our bitmask-dev fork main branch.

**Comment by zariye** [comment]
https://github.com/pixelated/pixelated-user-agent/tree/key_expiry

**Comment by zariye** [comment]
@bwagnerr 
> Do you think we should merge these changes to our forks develop so we can test the user agent properly with this on master?
yes!! i think we really should start making decisions here!

> Also we don't have a fork of gnupg, in that case we need start a fork of it (and use that fork on the UA), so we can merge the changes and review/use them
same answers, lets do it!

Actually I do not really like the idea of having forks generally, because you have to take care for them etc. but I think we should set us a day x (if day x is not already over) and just do things in order to getting forward with our process. Me personally started getting really confused with this ticket and where stuff is happening etc., and I do not think it will get less if we wait to long.

**Comment by tuliocasagrande** [comment]
We finally were able to reproduce an expired key scenario and used the **key_expiry** branch to regenerate the key. 🎉 

Further tests:
- [x] the key is renewed
- [x] the new key is signed by the old one
- [x] the user can read old emails
- [x] the user can read new emails
- [ ] the new key is sent attached on next emails (once for each contact)
- [ ] debug why a new welcome email is received after key change

**Comment by thaissiqueira** [comment]
Merged on master.
Further work will happen on issues #894 and #895.

**Comment by tuliocasagrande** [comment]
We've fixed the problem with the welcome mail being sent after regenerate key and pushed to bitmask-dev, feat_fetch_remote_on_expiry branch.

We're going to test it tomorrow.

**Comment by varac** [comment]
As a reminder, my wazokazi key is also still expired. I'm up for QAing if needed, but as it's a one-shot only operation, I'll hold on  until you give me green light.

**Comment by tayanefernandes** [comment]
We fixed the problem with the upload key failure, now it is going to attempt to upload again on the subsequent logins. This is a temporary solution because the fetch remote can increase the
login time, specially with multiple users. We can discuss with the team to think in a better solution.

Besides that, we still have some pending issues:

1. The first email after key renewal was not signed #894
2. When a user with lots of emails renew their keys, it's taking a long time to login. It seems it's encrypting all the emails with the new key, but we couldn't confirm that.
3. If a user tries to send an email to another user with an expired key, the interface puts the encrypted padlock (for instance varac expired key on the image), which is misleading for the user and the message is not sent. Instead of that, we could put the non-encrypted padlock and send the message unencrypted. This way we can avoid the user being unreachable during the expired key.

<img width="265" alt="screen shot 2017-01-10 at 3 56 33 pm" src="https://cloud.githubusercontent.com/assets/10910322/21818310/6dcfc7f6-d74e-11e6-915d-804ff13849cb.png">
